src/runtime/mutex.ts

```diff
@@ -0,0 +1,65 @@
+/**
+ * Running a Releaser releases the acquired mutex.
+ */
+export interface Releaser {
+  (): void;
+}
+
+/**
+ * A simmple Mutex to gate access to critical async code
+ * sections that should not execute concurrently.
+ *
+ * Sample usage:
+ * 
+ * ```
+ *   class SampleClass {
+ *     private readonly mutex = new Mutex();
+ *
+ *     async instantiate() {
+ *       const release = await mutex.acquire();
+ *       try {
+ *         // Protected section with async execution.
+ *       } finally {
+ *         release();
+ *       }
+ *     }
+ *   }
+ */
+
+export class Mutex {
+  private next: Promise<void> = Promise.resolve();
+  private depth = 0; // tracks the number of blocked executions on this lock.
+
+  /**
+   * @return true if the mutex is already acquired.
+   */
+  get locked(): boolean {
+    return this.depth !== 0;
+  }
+
+  /**
+   * Call acquire and await it to lock the critical section for the Mutex.
+   *
+   * @return A Releaser which resolves to a function which releases the Mutex.
+   */
+  async acquire(): Promise<Releaser> {
+
+    let release: Releaser;
+
+    const current: Promise<Releaser> = this.next.then(() => {
```

Comment:
Variable `this.next` starts out resolved, it seems to me it's `this.previous`, i.e. it's a Promise from a previous acquire (or a no-op). That's in the scope of the then() clause. In the scope of the return value it's actually the `current` (top-most) Promise. I find it a bit confusing, but maybe there is another way to look at it.

Reply:
I guess it makes sense if you look at it from outside the acquire..   Agree that this is tricky magic.